crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! MemoryCache is a cache for the transaction execution which delays writes to the database until
+//! transaction results are certified (i.e. they appear in a certified checkpoint, or an effects cert
+//! is observed by a fullnode). The cache also stores committed data in memory in order to serve
+//! future reads without hitting the database.
+//!
+//! For storing uncommitted transaction outputs, we cannot evict the data at all until it is written
+//! to disk. Committed data not only can be evicted, but it is also unbounded (imagine a stream of
+//! transactions that keep splitting a coin into smaller coins).
+//!
+//! We also want to be able to support negative cache hits (i.e. the case where we can determine an
+//! object does not exist without hitting the database).
+//!
+//! To achieve both of these goals, we split the cache data into two pieces, a dirty set and a cached
+//! set. The dirty set has no automatic evictions, data is only removed after being committed. The
+//! cached set is in a bounded-sized cache with automatic evictions. In order to support negative
+//! cache hits, we treat the two halves of the cache as FIFO queue. Newly written (dirty) versions are
+//! inserted to one end of the dirty queue. As versions are committed to disk, they are
+//! removed from the other end of the dirty queue and inserted into the cache queue. The cache queue
+//! is truncated if it exceeds its maximum size, by removing the oldest versions.
+//!
+//! This gives us the property that the sequence of versions in the dirty and cached queues are the
+//! most recent versions of the object, i.e. there can be no "gaps". This allows for the following:
+//!
+//!   - Negative cache hits: If the queried version is not in the cache, but is higher than the smallest
+//!     version in the queue, it does not exist in the db either.
+//!   - Bounded reads: When reading the most recent version that is <= some version bound, we can
+//!     correctly satisfy this query from the cache, or determine that we must go to the db.
```

Comment:
it should actually be "smallest (least) version in the cached queue".

I think this is all more confusing than it should be. the basic idea is that we keep the N most recent versions of an object in memory. If we had infinite memory, we would just have a single queue for each object, and it would all be very clear.

but, we don't have infinite memory, and therefore we need to split the queue into two pieces, a dirty side and a non-dirty side. The two sides require different strategies for bounding their size. The dirty side can only be bounded by slowing down execution / speeding up commits, while the non-dirty side can be bounded by automatic evictions. So it all ends up looking a lot more confusing than it should be.